contracts/vaults/Deposit.sol

```diff
@@ -49,37 +50,35 @@ contract Deposit is Base {
         require(_amount >= MINIMUM_DEPOSIT, "HVE27");
         //clear withdraw request
         withdrawEnableStartTime[_pid][msg.sender] = 0;
-        uint256 transferredAmount = _deposit(_pid, _amount);
+        PoolInfo storage pool = poolInfos[_pid];
+        UserInfo storage user = userInfo[_pid][msg.sender];
+        _claimReward(_pid, user);
+        uint256 lpSupply = pool.balance;
+        uint256 balanceBefore = pool.lpToken.balanceOf(address(this));
+        pool.lpToken.safeTransferFrom(address(msg.sender), address(this), _amount);
+        uint256 transferredAmount = pool.lpToken.balanceOf(address(this)) - balanceBefore;
+        pool.balance += transferredAmount;
+        uint256 userShares = transferredAmount;
+        // create new shares (and add to the user and the pool's shares) that are the relative part of the user's new deposit
+        // out of the pool's total supply, relative to the previous total shares in the pool
+        if (pool.totalShares > 0) {
+            userShares = pool.totalShares * transferredAmount / lpSupply;
+        }
+        user.shares += userShares;
+        pool.totalShares += userShares;
+        user.rewardDebt = user.shares * pool.rewardPerShare / 1e12;
         emit Deposit(msg.sender, _pid, _amount, transferredAmount);
     }
 
-    function _deposit(uint256 _pid, uint256 _amount) internal nonReentrant returns(uint256 _transferredAmount) {
+    function _claimReward(uint256 _pid, UserInfo memory _user) internal nonReentrant {
```

Comment:
what about moving the nonReentrant decorator to `deposit` and `withdraw` instead of hiding it here?